include/revng/Pipeline/Context.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <initializer_list>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Casting.h"
+
+#include "revng/Pipeline/KindsRegistry.h"
+#include "revng/Pipeline/SavableObject.h"
+
+namespace Pipeline {
+
+/// a simple pair that holds a savable object a name meant to describe a object
+/// which lifetime must be than the context.
+class NamedGlobalReference {
+private:
+  llvm::StringRef Name;
+  SavableObjectBase *Obj;
+
+public:
+  NamedGlobalReference(llvm::StringRef Name, SavableObjectBase &Obj) :
+    Name(Name), Obj(&Obj) {}
+
+  llvm::StringRef name() const { return Name; }
+
+  SavableObjectBase &value() const { return *Obj; }
+};
+
+/// A class that contains every object that has a lifetime longer
+/// than a pipeline.
+///
+/// This includes the kinds and the NamedGlobals that will be available in that
+/// context
+class Context {
+private:
+  llvm::StringMap<SavableObjectBase *> Globals;
+  KindsRegistry TheKindRegistry;
+
+public:
+  explicit Context(KindsRegistry registry) :
+    TheKindRegistry(std::move(registry)) {}
+  Context();
+  Context(llvm::ArrayRef<NamedGlobalReference> Globals, KindsRegistry registry);
+  Context(llvm::ArrayRef<NamedGlobalReference> Globals);
+  Context(const Context &) = default;
+  Context(Context &&) = default;
+  Context &operator=(const Context &) = default;
+  Context &operator=(Context &&) = default;
+  ~Context() = default;
+
+  template<typename T>
+  llvm::Expected<T *> getGlobal(llvm::StringRef Name) const {
+    auto Iter = Globals.find(Name);
+    if (Iter == Globals.end())
+      return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                     "pipeline loader context did not "
+                                     "contained object %s",
+                                     Name.str().c_str());
+    auto *Casted = llvm::dyn_cast<T>(Iter->second);
+    if (Casted == nullptr)
+      return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                     "pipeline loader was requested to cast %s "
+                                     "to the wrong type",
+                                     Name.str().c_str());
+
+    return Casted;
+  }
+
+  const KindsRegistry &getKindsRegistry() const { return TheKindRegistry; }
```

Comment:
So, let me understand, we can have completely distinct kind registries?
A kind can decide to what registry get attached? How does that work?

What do you think of getting in a "radix" kind as opposed to a `KindRegistry`? That might enable us to preserve a similar feature set, but have a single registry.

Reply:
when a context is created by default it copies the global kind registry which contains all of them, toh for testing purposes one my wish that the registry is not polluted with random kinds, so it can be overwritten, in the revng pipeline the global default one is used.

we can't have a radix kind that parents all of them, kinds in a tree must share the same granularity.

Reply:
Mmh, if tests don't want to be polluted... Can't they just not link to libraries they don't depend on?
Also, is it really a problem if the KindsRegistry is polluted?
It's quite strange to have magic auto-registering global variables, but then having an alternative way to do the same thing.

If you want to keep a copy of `KindsRegistry` in `Context`, keep one constructor (plus copy/move) and for the others make `static` factory methods with explicative names and documenting use cases.